### ISSUE TITLE
Discussion - input properties that get modified within charts

### ISSUE THREAD

**Issue by kum-deepak** [dc-js/dc.js#1739]
While working on #1736, I have come across a few properties which are modified within the `dc` code. For example `_colors` (linked to `.colors()`), sets color scale. This can be explicitly modified by calls like `.calculateColorDomain()` or `.colorDomain()`. Even calls like `.getColor()` may modify it by altering the domain implicitly.

I see the following options:

- We do not include these properties in conf.
- We defensively copy these values upon assignment.
- We explicitly mark such properties and expect users to pass new instances during configuring. Users looking to reuse configuration across charts will need to be careful.

Open for ideas :smile:

**Comment by gordonwoodhull** [comment]
This and #1735 property side effects indicate to me that we will have a difficult time making dc.js properties "pure" - it's a messy system which wasn't designed with such principles.

A big one which is relevant here is the X/Y scale - the user usually sets the scale (although it is sometimes defaulted), then dc.js sets the range on the user-provided scale, and often the domain too.

In one system I built on top of dc.js ([dcplot.js](https://github.com/att/dcplot.js)) I formalized stages of initialization - 

1. take user settings
2. default things that weren't set by user
3. infer missing properties from existing ones
4. check for required properties that still aren't set, and unknown properties
5. finally, generate

This worked pretty well but I don't know if it would work here.
